### PR TITLE
csi: validate the secretnames in CSI spec against NameIsDNSSubdomain

### PR DESCRIPTION
**Pull request by humblec**
csi: validate the secretnames in the CSI spec against NameIsDNSSubdomain
    
    At present the CSI spec secret name validation for ControllerPublish,
    ControllerExpand, NodePublish secrets are performed against
    ValidateDNS1123Label() and it causes the secret name validation
    inside the CSI spec to go wrong if the secret name is more than 63 chars.
    
    Kubernetes allow the secret object name to be on `DNS SubDomainName`
    and having a secret name length between 0-253 is correct/valid. So the CSI
    spec validation also has to be performed accordingly.
    
    This commit address this issue in validation for above mentioned funcs.

  
This PR also covers below items
- [x] Field name error report from `spec.csi.name` to `spec.csi.ControllerPublishSecretRef.name`, `spec.csi.ControllerExpandSecretRef.name`, `spec.csi.NodePublishSecretRef.name` ..etc
- [x] Unit tests for the validation `<=253 chars` and `>253 chars` for secretRefName in csi volume source

Fixes # https://github.com/kubernetes-csi/external-provisioner/issues/708

Additional Info;

Some of the ground work for this PR has been handled via below:

#108418 
#108391
https://github.com/kubernetes/kubernetes/pull/108420

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

/kind bug

-->
```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -1564,36 +1564,44 @@ func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldP
 		if len(csi.ControllerPublishSecretRef.Name) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath.Child("controllerPublishSecretRef", "name"), ""))
 		} else {
-			allErrs = append(allErrs, ValidateDNS1123Label(csi.ControllerPublishSecretRef.Name, fldPath.Child("name"))...)
+			for _, msg := range ValidateSecretName(csi.ControllerPublishSecretRef.Name, false) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("controllerPublishSecretRef"), csi.ControllerPublishSecretRef.Name, msg))
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:1568]: The invalid field is `ControllerPublishSecretRef.Name`, therefore `fldPath.Child("controllerPublishSecretRef", "name")`

(this applies to all similar cases below)

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -1564,36 +1564,44 @@ func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldP
...
+			}
 		}
 		if len(csi.ControllerPublishSecretRef.Namespace) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath.Child("controllerPublishSecretRef", "namespace"), ""))
 		} else {
 			allErrs = append(allErrs, ValidateDNS1123Label(csi.ControllerPublishSecretRef.Namespace, fldPath.Child("namespace"))...)
+
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:1575]: Why empty space?

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (humblec) [reply pkg/apis/core/validation/validation.go:1568]: ok.. sure.. thanks.

**Comment by jsafrane** [comment]
/label api-review
/priority important-soon
This is a clear bugfix and a good candidate to cherry-pick to older releases.

**Comment by humblec** [comment]
Cc @liggitt

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -1564,7 +1564,9 @@ func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldP
 		if len(csi.ControllerPublishSecretRef.Name) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath.Child("controllerPublishSecretRef", "name"), ""))
 		} else {
-			allErrs = append(allErrs, ValidateDNS1123Label(csi.ControllerPublishSecretRef.Name, fldPath.Child("name"))...)
+			for _, msg := range ValidateSecretName(csi.ControllerPublishSecretRef.Name, false) {
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:1567]: is this relaxed validation still within CSI spec?

assuming it is, and we want this change long-term, we cannot relax validation in a single release, or we'll allow persisting data the previous version of the API server considers invalid, breaking updates via that older API server. relaxing the validation has to be done over two releases, first tolerating the relaxed data in update requests of objects already containing relaxed data, then in the next release, permitting the relaxed validation on all create and update requests.

Comment (jsafrane) [reply pkg/apis/core/validation/validation.go:1567]: > is this relaxed validation still within CSI spec?

Yes, CSI does not describe any Kubernetes object. A CSI driver gets content of a Secret, not its name.

Comment (liggitt) [reply pkg/apis/core/validation/validation.go:1567]: the process of relaxing validation is similar to that of permitting a new enum value in an existing field, described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

**Comment by liggitt** [comment]
> This is a clear bugfix and a good candidate to cherry-pick to older releases.

I disagree, it's an API validation change we should not backport. Using secret names that fit within the current validation is a reasonable workaround until the relaxed validation is released.

File: pkg/apis/core/validation/validation.go (hunk 3 above)
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:1567]: is this value propagated to any other API fields that also have tight validation, or combined with other values and saved in API fields in ways that would be broken by allowing longer values for the secret name here?

Comment (humblec) [reply pkg/apis/core/validation/validation.go:1567]: afaict, only csi volume source ( which can be part of inline volumes) and pv.spec.csi (for pvc claims) are the affected areas. however I would let Jan to share his thoughts as well.

**Comment by humblec** [comment]
> > This is a clear bugfix and a good candidate to cherry-pick to older releases.
> 
> I disagree, it's an API validation change we should not backport. Using secret names that fit within the current validation is a reasonable workaround until the relaxed validation is released.

Sure, agreed :+1:

**Comment by jsafrane** [comment]
Looking how it's done for `ReadWriteOncePod` access mode, it makes more sense to move `opts` creation here:  https://github.com/kubernetes/kubernetes/blob/92263ee0dd2588978dd2d99a952a1d463e3a88a3/pkg/apis/core/validation/validation.go#L1660

`validation.go` then should only obey and not update `opts` and pass it where it needs them.

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -1926,7 +1975,13 @@ func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName stri
...
 		} else {
 			numVolumes++
-			allErrs = append(allErrs, validateCSIPersistentVolumeSource(pvSpec.CSI, fldPath.Child("csi"))...)
+			// if the opts carry flag for allowing DNS subDomain secret names, set the callee flag.
+			if opts.AllowDNS1123SubDomainSecrets {
+				allErrs = append(allErrs, validateCSIPersistentVolumeSource(pvSpec.CSI, true, fldPath.Child("csi"))...)
+
+			} else {
+				allErrs = append(allErrs, validateCSIPersistentVolumeSource(pvSpec.CSI, false, fldPath.Child("csi"))...)
+			}
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:1984]: Why `if`?
`validateCSIPersistentVolumeSource(pvSpec.CSI, opts.AllowDNS1123SubDomainSecrets, ...`

File: pkg/apis/core/validation/validation.go [hunk 5]
```diff
@@ -1918,6 +1966,7 @@ func ValidatePersistentVolumeSpec(pvSpec *core.PersistentVolumeSpec, pvName stri
 		} else {
 			numVolumes++
 			allErrs = append(allErrs, validateStorageOSPersistentVolumeSource(pvSpec.StorageOS, fldPath.Child("storageos"))...)
+
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:1969]: Why empty line here?

File: pkg/apis/core/validation/validation.go [hunk 6]
```diff
@@ -1532,12 +1532,22 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
...
 // validatePVSecretReference check whether provided SecretReference object is valid in terms of secret name and namespace.
 
-func validatePVSecretReference(secretRef *core.SecretReference, fldPath *field.Path) field.ErrorList {
+func validatePVSecretReference(secretRef *core.SecretReference, allowDNSSubDomainFormat bool, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
 	if len(secretRef.Name) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
 	} else {
-		allErrs = append(allErrs, ValidateDNS1123Label(secretRef.Name, fldPath.Child("name"))...)
+		if !allowDNSSubDomainFormat {
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:1540]: Nit: don't negate `allowDNSSubDomainFormat`, it's hard to read. Switch `then` / `else` branches below.

File: pkg/apis/core/validation/validation_test.go [hunk 7]
```diff
@@ -2909,7 +2909,7 @@ func TestValidateCSIPersistentVolumeSource(t *testing.T) {
 	}
 
 	for i, tc := range testCases {
-		errs := validateCSIPersistentVolumeSource(tc.csi, field.NewPath("field"))
+		errs := validateCSIPersistentVolumeSource(tc.csi, false, field.NewPath("field"))
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation_test.go:2912]: Please add some unit tests with `true` to check the new behavior.

File: pkg/apis/core/validation/validation.go (hunk 6 above)
Comment (humblec) [reply pkg/apis/core/validation/validation.go]: sure, done

File: pkg/apis/core/validation/validation_test.go (hunk 7 above)
Comment (humblec) [reply pkg/apis/core/validation/validation_test.go]: Done. thanks.

**Comment by jsafrane** [comment]
/lgtm
@liggitt PTAL or assign an API reviewer

File: pkg/apis/core/validation/validation.go [hunk 8]
```diff
@@ -1669,6 +1681,42 @@ func ValidationOptionsForPersistentVolume(pv, oldPv *core.PersistentVolume) Pers
 		// If the old object allowed "ReadWriteOncePod", continue to allow it in the new object
 		opts.AllowReadWriteOncePod = true
 	}
+	// if its CSI spec and secretName was DNS1035Label format for oldPV
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:1684]: These blocks are hard to review and vulnerable to copy/paste errors, and is doing unnecessary evaluation if one of the earlier blocks already set opts.AllowDNS1123SubDomainSecrets to `true`. Can this be structured something like this instead:

```go
if oldCSI := oldPv.Spec.CSI; oldCSI != nil {
  opts.AllowDNS1123SubDomainSecrets = 
    secretRefRequiresSubdomainSecretName(oldCSI.ControllerExpandSecretRef) ||
    secretRefRequiresSubdomainSecretName(oldCSI.ControllerPublishSecretRef) ||
    secretRefRequiresSubdomainSecretName(oldCSI.NodeStageSecretRef) ||
    secretRefRequiresSubdomainSecretName(oldCSI.NodePublishSecretRef)    
  }
}
...

func secretRefRequiresSubdomainSecretName(secretRef *core.SecretReference) bool {
  // ref and name were specified and name didn't fit within label validation
  return secretRef != nil && len(secretRef.Name) > 0 && len(validation.IsDNS1123Label(secretRef.Name)) > 0
}
```

Comment (humblec) [reply pkg/apis/core/validation/validation.go]: agreed.  rearranged to follow above format :+1:

File: pkg/apis/core/validation/validation.go [hunk 9]
```diff
@@ -1532,12 +1532,22 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
 
 // validatePVSecretReference check whether provided SecretReference object is valid in terms of secret name and namespace.
 
-func validatePVSecretReference(secretRef *core.SecretReference, fldPath *field.Path) field.ErrorList {
+func validatePVSecretReference(secretRef *core.SecretReference, allowDNSSubDomainFormat bool, fldPath *field.Path) field.ErrorList {
```
[the remark this answers is not in the document]
Comment (humblec) [reply pkg/apis/core/validation/validation.go]: Sure, renamed the arg

File: pkg/apis/core/validation/validation.go [hunk 10]
```diff
@@ -1644,7 +1647,8 @@ var allowedPVCTemplateObjectMetaFields = map[string]bool{
 // PersistentVolumeSpecValidationOptions contains the different settings for PeristentVolume validation
 type PersistentVolumeSpecValidationOptions struct {
 	// Allow spec to contain the "ReadWiteOncePod" access mode
-	AllowReadWriteOncePod bool
+	AllowReadWriteOncePod       bool
+	allowDNSSubDomainSecretName bool
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:1651]: since this struct is an arg to a public method, this field should be public as well

Comment (liggitt) [reply pkg/apis/core/validation/validation.go:1651]: and add godoc for the field

File: pkg/apis/core/validation/validation_test.go [hunk 11]
```diff
@@ -20530,7 +20609,7 @@ func TestValidatePVSecretReference(t *testing.T) {
 			name:          "invalid secret ref name",
 			args:          args{&core.SecretReference{Name: "$%^&*#", Namespace: "default"}, rootFld},
 			expectError:   true,
-			expectedError: "name.name: Invalid value: \"$%^&*#\": " + dnsLabelErrMsg,
+			expectedError: "name: Invalid value: \"$%^&*#\": " + dnsLabelErrMsg,
```
Comment (liggitt) [inline pkg/apis/core/validation/validation_test.go:20612]: revert this change, if we're saying the root field is "name", then the name attribute of the ref is `name.name`

Comment (humblec) [reply pkg/apis/core/validation/validation_test.go]: sure, done.

File: pkg/apis/core/validation/validation.go (hunk 10 above)
Comment (humblec) [reply pkg/apis/core/validation/validation.go]: Done.. @liggitt .. thanks..

File: pkg/apis/core/validation/validation.go [hunk 12]
```diff
@@ -1644,7 +1647,9 @@ var allowedPVCTemplateObjectMetaFields = map[string]bool{
 // PersistentVolumeSpecValidationOptions contains the different settings for PeristentVolume validation
 type PersistentVolumeSpecValidationOptions struct {
 	// Allow spec to contain the "ReadWiteOncePod" access mode
-	AllowReadWriteOncePod bool
+	AllowReadWriteOncePod       bool
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:1650]: I bet gofmt will want this line back the way it was

Comment (humblec) [reply pkg/apis/core/validation/validation.go]: @liggitt true :), Done.

**Comment by humblec** [comment]
Thanks @liggit and @jsafrane :+1:
